### PR TITLE
[FLINK-12715][hive] Hive-1.2.1 build is broken

### PR DESCRIPTION
**Pull request by lirui-apache**
## What is the purpose of the change

To fix hive-1.2.1 build. The build was broken because some util method used in FLINK-12568 is not compatible in Hive-2.3.4 and Hive-1.2.1.

## Brief change log

  - Added shim for `FileUtils.moveToTrash`.
  - Added test to cover it.

## Verifying this change

Added new test and tested for both 2.3.4 and 1.2.1 (manually).

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

### CODE REVIEW

**Comment by lirui-apache** [comment]
cc @zjuwangg @bowenli86 @xuefuz

**Comment by xuefuz** [comment]
LGTM

**Comment by bowenli86** [comment]
I wonder that, since`moveToTrash()` is mainly a util, should we move it to a separate util shim class like `HiveUtilShim`? The current `HiveShim` class, as its javadoc says, is mainly for Hive Metastore. We probably should rename it to `HiveClientShim`?

LGTM otherwise.

**Comment by xuefuz** [comment]
> I wonder that, since`moveToTrash()` is mainly a util, should we move it to a separate util shim class like `HiveUtilShim`? The current `HiveShim` class, as its javadoc says, is mainly for Hive Metastore. We probably should rename it to `HiveClientShim`?
> 
> LGTM otherwise.

HiveShim is quite generic in naming. It's a shim, so in theory anything can be put there. Maybe we just need to update the javadoc to reflect the new addition.

Having another shim for utils might be an overkill for now. However, we can keep an eye on this and see how we organize better when this layer gets too heavy.

**Comment by lirui-apache** [comment]
I also prefer to keep HiveShim generic. Actually APIs in HiveShim like `getViews` and `getFunction` take a `IMetaStoreClient` as parameter. So in that sense it's generic in nature.
@bowenli86 what do you think?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV1.java [hunk 1]
```diff
@@ -79,4 +85,14 @@ public Function getFunction(IMetaStoreClient client, String dbName, String funct
 			throw e;
 		}
 	}
+
+	@Override
+	public boolean moveToTrash(FileSystem fs, Path path, Configuration conf, boolean purge) throws IOException {
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV1.java]: how is the "purge" flag taking effects?

Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV1.java]: Hive-1.2.1 doesn't support auto purge for insert overwrite, see [HIVE-15880](https://issues.apache.org/jira/browse/HIVE-15880). So it doesn't have any effect for HiveShimV1.

Comment (bowenli86) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV1.java]: I see, can we add a comment to the interface's param explaining the flag only works for hive 2 (or 2.1.x, or any version it works)?

**Comment by bowenli86** [comment]
Thanks @lirui-apache for the PR and review from @xuefuz . 

LGTM. Merging

**Pull request closed without merging**